### ISSUE TITLE
[UX][PS] Replace the current token browser in core with the Fast Token Browser implementation from contrib

### ISSUE THREAD

**Issue by herbdool** [backdrop/backdrop-issues#6143]
## Description of the need

I think https://github.com/backdrop-contrib/fast_token_browser is an improvement over the core version. It loads faster, has a more intuitive UI, and will be easier to make accessible to use via the keyboard (maybe it does already).

## Proposed solution

Put FTB into core.

## Additional information

FTB loads the tokens via ajax which makes it faster when there are a lot of tokens. And it also changes the way tokens are inserted. First the person selects the token - it shows up as selected. Then they click on / select the text field. It then gets inserted. Core has a clunkier approach which requires the unintuitive approach of selecting the field first then finding the token.

Also the modal window can be resized and moved.

![2023-06-12 11 40 14 bd1 lndo site 8995771bfaa3](https://github.com/backdrop/backdrop-issues/assets/1071483/e5ca245c-048d-4c5b-8fca-d5270c4c940a)

**Comment by klonos** [comment]
I really like this suggestion 👍🏼 

...but for the sake of playing the devil's advocate:
- The contrib module counts only 65 active installations at the moment, which given the ~2500 Backdrop sites reporting stats makes it less than 3%, which is too low. But then again, this might be one of the cases where if people knew about the existence of this module, they'd use it more 🤷🏼 
- Not sure whether the way the tokens are being inserted in FTB is ideal 🤔 ...the token seems to always be appended to the end of the existing text in the target text field/area, which makes it really cumbersome and annoying when the intention is to add things in the middle or start of the text.

**Comment by yorkshire-pudding** [comment]
Please see https://github.com/backdrop-contrib/fast_token_browser/issues/3 - I do think the UX should be improved before it is merged into core.

As @klonos is the devil's advocate, I should probably be the angel's advocate and say that core breaks if you have a lot of tokens which this fixes.  If we can get the UX improved, then it is not a new feature per se but a replacement of an existing feature.

**Comment by klonos** [comment]
Yes, if we managed to address that UX issue, I'd be happy for this to get into core. The way that FTB works (loading tokens via AJAX as needed) sounds like how things should be working to begin with.

**Comment by klonos** [comment]
...I've added the `[PS]` (Performance & Scalability) "tag" in the title of this issue, as per https://docs.backdropcms.org/documentation/how-to-use-the-core-issue-queue, based on the fact that the implementation of loading tokens in FTB via AJAX is a performance improvement.

**Comment by yorkshire-pudding** [comment]
@klonos - the latest release with the changes from @argiepiano , I believe, have addressed your concerns.

**Comment by klonos** [comment]
Thanks @yorkshire-pudding 🙏🏼 and thank you @argiepiano 🙏🏼 🙏🏼 

I've given the FTB module another spin, and the UX I had concerns for has solved, but there is still a small bug for a follow-up:
1. place the cursor in the middle of the existing text in a token-enabled field
2. trigger the token browser
3. click on a random token -> the token gets added to the place where the cursor was placed in step 1 👍🏼 
4. click on another token -> the token gets added to the start of the text 👎🏼 ...I'd expect it to be added right after the token added in step 3

But perhaps we can leave that to solved in the PR for merging the functionality into core. Anyone has the time/energy to work on that?

Other UX/consistency things that I would like us to consider when in core:
- Rename the 1st column in the token browser window from "Name" to "Token"
- Merge the human-readable name of the token and the token itself into the 1st column.
- The token should be rendered via `theme_label_machine_name()`, and not be clickable.
- A new column should be added to the table, with "Place token" buttons that are styled the same as the regular "operations" buttons we have in other tables in the admin UI (instead of the links used now to add the tokens)
- The machine names for the top-level tokens should be changed from `node`, `current-date`, `current-page` to `[node:...]`, `[current-date:...]`, `[current-page:...]` etc. instead.
- Add zebra stripping to the table rows.
- Add hover-over CSS color to the table rows.
- Add scroll within the token browser modal. Currently, it keeps being getting longer and longer as you expand more token categories, and at some point the tokens you are after are past the point where the target field is visible. You cannot drag and drop the token browser window, since it cannot be dragged above the top of the page.
- Adjust the z-index of the token browser window, so that it is below the admin bar and any menu item in the admin bar.
- Add the name of the target field to the title of the browser window. So instead of the static "Token Browser" title, there should be a "Add tokens to field xyz".

That's it for now. Curious to see what people think about the above suggestions, and I'd be happy to do another round of review and make further UX suggestions once we have a PR sandbox to play with.

**Comment by argiepiano** [comment]
@klonos said:

> click on a random token -> the token gets added to the place where the cursor was placed in step 1 👍🏼
click on another token -> the token gets added to the start of the text 👎🏼 ...I'd expect it to be added right after the token added in step 3

But but but... the issue was to mirror the behavior of the core browser, and this is exactly what the core browser does!

**Comment by argiepiano** [comment]
@klonos , thanks for the "laundry list" of changes. I was going to take a stab at this, but this list is a rather daunting list now - it forshadows an issue that will be in the queue for years, so perhaps it's best to leave this as contrib? Or perhaps others feel that this list is not a "sine que non"? 

Anyway, besides those changes, there are a few important code-wise changes to make to the JS. The JS is very readable and clean, but unfortunately doesn't use namespacing for variables or functions - instead it stores things as global `window` variables that could be accidentally overwritten, making the code less than safe. Even if this remains in contrib, this needs to be fixed. @yorkshire-pudding I'll submit another PR fixing some of those issues.

**Comment by yorkshire-pudding** [comment]
I'm a firm believer that we shouldn't let the desire for perfect stand in the way of incremental improvement (I actually that this should be a "Backdrop principle").

With the wonderful improvements already by @argiepiano and the suggested code namespacing refinement, perhaps this should just focus on anything important that is missing from this module  when compared to core, with anything else being separated for later improvement.

**Comment by klonos** [comment]
> ... I was going to take a stab at this, but this is a rather daunting list now - it forshadows an issue that will be in the queue for years ...

Apologies @argiepiano ...I didn't mean to deter you or anyone else from working on this. Please do not let my OCDs and extensive list of "ideal" get in the way - everything I listed in my last comment can be follow-ups (or I can file PRs agains any PR created by others here for consideration). I simply wanted to document these points, since I took the time to do a UX review. And since we do not have any other ticket to do that, here seemed like a relevant place to do it (it'd be silly to file a follow-up for an issue that doesn't even have a PR yet).

> I'm a firm believer that we shouldn't let the desire for perfect stand in the way of incremental improvement (I actually think that this should be a "Backdrop principle").

Yes! That ^^ 👍🏼

**Comment by quicksketch** [comment]
I think this would be a great 1.30.0 feature. Is anyone interested in being the advocate?

**Comment by herbdool** [comment]
I think refactoring the JS might be a blocker https://github.com/backdrop-contrib/fast_token_browser/issues/19.

**Comment by yorkshire-pudding** [comment]
> I think refactoring the JS might be a blocker [backdrop-contrib/fast_token_browser#19](https://github.com/backdrop-contrib/fast_token_browser/issues/19).

I love the Backdrop community. Yes, this was a blocker but within a few hours of this message, there were two PRs to fix it; thanks to both @herbdool and @argiepiano .

**Comment by argiepiano** [comment]
> I think refactoring the JS might be a blocker https://github.com/backdrop-contrib/fast_token_browser/issues/19.

That issue has been fixed.

**Comment by argiepiano** [comment]
Just to avoid duplicating efforts - I'll try a PR to move this to core. I'll let you know ASAP if this becomes too much at the moment, so someone else can take over (it's fun to work on Backdrop, but other duties are competing for my time too).

**Comment by herbdool** [comment]
@argiepiano give it a shot. There's no rush. I'm trying to fix a storm window IRL.

**Comment by argiepiano** [comment]
I have submitted an initial PR, WIP.  It moves the code from Fast Token Browser into core's `system.module`, `system.pages.inc`, `system.theme.inc` etc, replacing the current token browser completely. This is working on my local.

Aspects that need work:
- The flag `show_restricted` doesn't yet work (this was a problem of Fast TokenBrowser). This flag is used to hide certain tokens that are not supposed to be used, like `[node:tnid]` (translation id).
- The flag `click_insert` is not supported (yet). I have to figure out if this is needed, since the fast token tree approach is different. This flag basically disabled the ability to insert a token into a DOM field by clicking it on the token browser

I'm noticing that there are test failures - probably related to token UI tests. I'll check on this later today.

@yorkshire-pudding, if you fancy doing some testing, you can patch core with the PR and test the browser, for example by going to `admin/config/urls/path/patterns` and `admin/structure/types/manage/card`.

Also, you may want to enable modules like Token Help token tree that appears at `admin/reports/tokens`

**Comment by argiepiano** [comment]
I fixed the issue that prevented the sandbox from being built. It can be tested on the sandbox now.

**Comment by yorkshire-pudding** [comment]
Thanks @argiepiano. Fast work.
I wasn't aware of the flag 'show_restricted' or that it was a problem with FTB. I guess I've used the module on all my sites since 2022, so I've never known anything different.

It all seems to work for me.  I'll hold off adding works for me label as you've indicated it needs more work but what is there looks great.

@herbdool - as you created this issue initially, would you be willing to be the 'Advocate' for it?  I'm already advocating for another issue.

**Comment by argiepiano** [comment]
I added the flags `'show_restricted`` and `'click_insert'`. This means that restricted tokens like `[node:tnid]` are not shown in the tokens opened by the "URL alias pattern" tab.

Also, the click_insert flag assures that Token Help shows a token tree without the clickable links to insert them. 

Ready for testing and review.

**Comment by yorkshire-pudding** [comment]
Thanks @argiepiano . I have tested the revised PR and can confirm it all still works.  Plus:

- I do not now see `[node:tnid]` in url alias page. also don't see on token block.  I don't if or where it actually is visible but that is not relevant.
- I do not see an insert link in Token Help - just the text of the token (I wondered about this for ages in the module)

Therefore all works.

**Comment by herbdool** [comment]
I can advocate for it.

**Comment by herbdool** [comment]
Code looks good. Though I'm wondering @argiepiano how much work it would be to first add the flags `show_restricted` and `click_insert` to the contrib module so we can review those separately from the rest of the code. Then it'll be clear that it's just moving code in.

I suppose we don't want to bring along the git history since the code is getting spread around the `system` module. So I'm okay with that.

**Comment by argiepiano** [comment]
@herbdool thanks. I worked on adding this functionality to Fast Token Browser. As I was working on that I discovered an issue with the token cache that Fast Token Browser uses, when adding `show_restricted`. In a nutshell, the tokens for any given type are cached as they are retrieved (via ajax). 

The problem is that this cache doesn't take into account whether `show_restricted` is TRUE or FALSE. So, for example, if you open the "Content" tokens here (where show_restricted is FALSE): `admin/config/urls/path/patterns`, and then open the Token Help list of tokens (where show_restricted is TRUE), the latter will NOT show the restricted tokens, because it uses the cached tokens generated in the first list of tokens (where show_restricted was FALSE).

I think I know how to fix this. I'll give that a try next. Setting to NW.

**Comment by argiepiano** [comment]
PR has now been fixed. To check that it works:

- Open the sandbox for the PR
- Install and enable Token Help
- Clear caches (clear browser caches to be on the safe side as well)
- Visit `admin/config/urls/path/patterns`, open the "Content" token type,  and verify that `[node:tnid]` is not included 👍🏽 
- Visit `admin/reports/tokens` and verify that `[node:tnid]` is listed there (before this fix, it wasn't, as it was using the cached version of the token output with `show_restricted` = FALSE)

The way I solved this is by using the variable `$show_restricted` when building the hash that is used as the cache ID. This way, Backdrop creates two different cache entries for the same token types, when the token output includes or excludes the restricted ones.

**Comment by yorkshire-pudding** [comment]
> The way I solved this is by using the variable `$show_restricted` when building the hash that is used as the cache ID. This way, Backdrop creates two different cache entries for the same token types, when the token output includes or excludes the restricted ones.

Clever solution

Tested and works for me.

**Comment by herbdool** [comment]
All looks good to me now. Thanks @argiepiano!

**Comment by quicksketch** [comment]
I tried out the sandbox and it works pretty good! I left a bunch of feedback: https://github.com/backdrop/backdrop/pull/4884#pullrequestreview-2396584152

I didn't review all the JavaScript, quite yet, but there are several things that could be addressed.

**Comment by argiepiano** [comment]
Ok, thanks @quicksketch. Will get to this over the weekend.

**Comment by argiepiano** [comment]
@quicksketch see my comments in the PR. Switching to needs test and needs review again.

**Comment by quicksketch** [comment]
Looking great @argiepiano! I left another comment on the PR.

**Comment by argiepiano** [comment]
PR rebased and modifies to address latest comments.

**Comment by quicksketch** [comment]
The latest changes (code nit-picks aside), make it so that if a token browser only contains a single token type, the token tree is automatically expanded. This happens on content type settings for example when setting a URL alias pattern:

![image](https://github.com/user-attachments/assets/e4cda06b-7642-48c7-9e82-2668bd369c98)

But if there are two or more token categories, the trees stay collapsed until clicked, like on the taxonomy vocabulary URL alias pattern:

![image](https://github.com/user-attachments/assets/decac50d-914d-486b-bd5e-15395c18c53d)

Works great! Thanks for making that change @argiepiano.

The key thing I see missing here now is handling upgrades on sites that had Fast Token Browser module installed previously. The old module name should be added to `backdrop_merged_modules()`, and we should add an update hook in `system.install` that uninstalls the old module.

This looks all-but-certain for 1.30.0, could you add `@since 1.30.0 Function added.` to the docblock of all the newly added functions? I also left another round of feedback on the PR: https://github.com/backdrop/backdrop/pull/4884#pullrequestreview-2399724077

**Comment by argiepiano** [comment]
@quicksketch all comments have been addressed. Please check and test.

**Comment by argiepiano** [comment]
Note: the coding standard test is showing a white space in a file that was not touched by this PR. It must be from another PR?

**Comment by quicksketch** [comment]
@argiepiano Yes, don't worry about changes to files unmodified by this PR. This might be another variation of the issue reported at #5848 .

**Comment by quicksketch** [comment]
This looks good to me! I'd love to get some additional manual testing on this. @herbdool, @yorkshire-pudding, (maybe @stpaultim?) could you give things a spin now that code changes have been addressed?

**Comment by herbdool** [comment]
Just tested the demo on a phone. On the url aliases patterns page.

When I open token browser the first time, it works well. But when I close it and reopen it, the I can't open sections. No arrow.

**Comment by argiepiano** [comment]
I'm seeing something similar, but the arrow are there. It has to do with the size of the modal. If I turn my phone sideways and move the modal, I can see the arrows. My guess is that this has to do with dialog issues, not with this PR.

**Comment by argiepiano** [comment]
I'm going to check the css and dialog settings, compared to the old token browser to fix this.

**Comment by argiepiano** [comment]
@herbdool I've fixed the mobile problem. It was a css issue.

BTW I found TONS of issues with the admin bar in my Android phone. The bar is not hamburguerized on some pages. I'll open an issue for this.

**Comment by yorkshire-pudding** [comment]
Thanks @argiepiano and @quicksketch . I have tested this on desktop and mobile and it is both looking good and working well.
Thanks.  I'm happy to put "works for me"  but I know Nate was looking for a few more manual testers so I'll leave for now.

**Comment by herbdool** [comment]
That now works better on mobile for me. Thanks

**Comment by herbdool** [comment]
I also tried it in the browser and it works well.

**Comment by quicksketch** [comment]
I have merged https://github.com/backdrop/backdrop/pull/4884 into 1.x for 1.30.x. Great work @argiepiano!! Thank you for taking the initiative on this one and converting Fast Token Browser from an adequate-but-outdated contrib module to a nice refactored and updated version for core.

Thank you @herbdool, @yorkshire-pudding, and @klonos for the discussion, testing, and code review! First big feature of 1.30.0. :tada:

**Comment by quicksketch** [comment]
We still need a change record for this. Here's a good example of when Date module was put in core: https://docs.backdropcms.org/change-records/date-date-popup-and-date-views-modules-now-included-in-core

**Comment by yorkshire-pudding** [comment]
I'm starting to draft the Change record.

**Comment by yorkshire-pudding** [comment]
@quicksketch - following your explanation of how the `backdrop_modules_merged()` function works, I gave it a test so I could make sure the change record reflects what site admins can expect.

While it does prevent FTB from loading, use the new core browser, and show a clear(ish) message in the list modules page:
![image](https://github.com/user-attachments/assets/46c90e22-48b9-40ab-a23a-9af0342186a4)

It does not appear to actually disable the module in the system table or actually allow it to be disabled.  See Bee output below.

<details>
<summary>:bee: Bee output after patching</summary>

```
user@machine:~/projects/backdrop/bee/bee_dev/backdrop$ lando bee projects fast_token_browser
 Name          Fast Token Browser
 Type          Module
 Project       fast_token_browser
 Description   A more efficient browser for the Token module.
 Configure
 Package       User Interface
 Tags          Tokens, Administration, Theme Enhancements
 Version       1.x-1.6.2
 Path          modules/fast_token_browser
 Status        Enabled
 Requires
 Required by

user@machine:~/projects/backdrop/bee/bee_dev/backdrop$ lando bee dis fast_token_browser

 ✔  The 'Fast Token Browser' module was disabled.

user@machine:~/projects/backdrop/bee/bee_dev/backdrop$ lando bee pmu fast_token_browser

 ✘  The 'Fast Token Browser' module is still enabled and so cannot be uninstalled. Disable it and try again.

user@machine:~/projects/backdrop/bee/bee_dev/backdrop$ lando bee projects fast_token_browser
 Name          Fast Token Browser
 Type          Module
 Project       fast_token_browser
 Description   A more efficient browser for the Token module.
 Configure
 Package       User Interface
 Tags          Tokens, Administration, Theme Enhancements
 Version       1.x-1.6.2
 Path          modules/fast_token_browser
 Status        Enabled
 Requires
 Required by
```

</details>

I can get rid of it by deleting the module folder.  Do we need a separate ticket to improve this behaviour so that:

1. It actually disables the module; and 
2. It is clearer about what needs to be done to remove from the file system?

**Comment by yorkshire-pudding** [comment]
Further to the above, deleting from the file system leaves it in the `system` table:

```mysql
mysql> SELECT * FROM system WHERE `name` = 'fast_token_browser';
+------------------------------------------------------+--------------------+--------+-------+--------+-----------+----------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| filename                                             | name               | type   | owner | status | bootstrap | schema_version | weight | info
+------------------------------------------------------+--------------------+--------+-------+--------+-----------+----------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| modules/fast_token_browser/fast_token_browser.module | fast_token_browser | module |       |      1 |         0 |              0 |      0 | a:17:{s:4:"name";s:18:"Fast Token Browser";s:11:"description";s:46:"A more efficient browser for the Token module.";s:8:"backdrop";N;s:4:"type";s:6:"module";s:11:"maintainers";a:1:{s:17:"yorkshire-pudding";s:12:"Martin Price";}s:18:"maintainers_wanted";b:1;s:7:"package";s:14:"User Interface";s:4:"tags";a:3:{i:0;s:6:"Tokens";i:1;s:14:"Administration";i:2;s:18:"Theme Enhancements";}s:7:"project";s:18:"fast_token_browser";s:7:"version";s:9:"1.x-1.6.2";s:9:"timestamp";s:10:"1729070053";s:5:"mtime";i:1730456288;s:12:"dependencies";a:0:{}s:3:"php";s:5:"5.6.0";s:5:"files";a:0:{}s:9:"bootstrap";i:0;s:16:"merged_into_core";b:1;} |
+------------------------------------------------------+--------------------+--------+-------+--------+-----------+----------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

**Comment by yorkshire-pudding** [comment]
@quicksketch - I've created a separate issue to look at this: #6744

**Comment by argiepiano** [comment]
@yorkshire-pudding did you run the update? The update that comes with the PR actually removes the entry from the system table completely. If you don't delete the folder, a new entry may be created, but it will not be enabled when Backdrop creates that entry, and you won't be able to enable it again.

**Comment by yorkshire-pudding** [comment]
Hmm. For some reason, the update prompt didn't appear earlier.  I've re-tested and it does now appear and does the job.  I'll close that other ticket.

**Comment by yorkshire-pudding** [comment]
Updated Draft Change record - see https://github.com/backdrop/backdrop-issues/issues/6143#issuecomment-2446256905

**Comment by herbdool** [comment]
I've got a really small tweak. I added a new PR since this already merged, but felt it was too small to create a new issue (though I could if people think that's better).

All this does is add a pointer so it's obvious where to click to expand the token tree.

Before:

![Screenshot from 2024-11-01 09-52-06](https://github.com/user-attachments/assets/af3c9adf-c17a-4968-9959-222393a7d671)

After:

![Screenshot from 2024-11-01 09-51-41](https://github.com/user-attachments/assets/603c4fa8-896f-4217-9f43-42c4096382cb)
